schematools/contrib/django/factories.py

```diff
@@ -27,6 +27,212 @@
+    def fetch_relation_parts(cls, relation):
+        relation_parts = relation.split(":")
+        return relation_parts[:2]
+
+    @classmethod
+    def is_loose_relation(cls, relation, dataset, field):
+        """Determine if relation is loose or not."""
+
+        related_dataset_id, related_table_id = cls.fetch_relation_parts(relation)
+        related_dataset = dataset.get_dataset_schema(related_dataset_id)
+        related_table = related_dataset.get_table_by_id(related_table_id)
+
+        # Short-circuit for non-temporal or on-the-fly (through or nested) schemas
+        if (
+            not related_table.is_temporal
+            or field._parent_table.is_through_table
+            or field._parent_table.is_nested_table
+        ):
+            return False
+
+        # So, target-side of relation is temporal
+        # Determine fieldnames used for temporal
+        sequence_identifier = related_table.temporal["identifier"]
+        identifier = related_dataset.identifier
+
+        # If temporal, this implicates that the type is not a scalar
+        # but needs to be more complex (object) or array_of_objects
+        if field.type in set(["string", "integer"]) or field.is_array_of_scalars:
+            return True
+
+        sequence_field = related_table.get_field_by_id(sequence_identifier)
+        identifier_field = related_table.get_field_by_id(identifier)
+        if sequence_field is None or identifier_field is None:
+            raise ValueError(f"Cannot find temporal fields of table {related_table.id}")
+
+        if field.is_array_of_objects:
+            properties = field.items["properties"]
+        elif field.is_object:
+            properties = field["properties"]
+        else:
+            raise ValueError("Relations should have string/array/object type")
+
+        source_type_set = set(
+            [(prop_name, prop_val["type"]) for prop_name, prop_val in properties.items()]
+        )
+        destination_type_set = set(
+            [
+                (sequence_field.name, sequence_field.type),
+                (identifier_field.name, identifier_field.type),
+            ]
+        )
+
+        return source_type_set != destination_type_set
+
+    @classmethod
+    def fetch_maker(cls, dataset, field):
+        # determine type of relation (FKLoose, FK, M2M, LooseM2M)
+        if field.relation:
+            if cls.is_loose_relation(field.relation, dataset, field):
+                return LooseFKRelationMaker
+            else:
+                return FKRelationMaker
+        elif field.nm_relation:
+            if cls.is_loose_relation(field.nm_relation, dataset, field):
+                return LooseM2MRelationMaker
+            else:
+                return M2MRelationMaker
+        else:
+            return None  # To signal this is not a relation
+
+    def _make_related_classname(self, relation):
+        related_dataset, related_table = [to_snake_case(part) for part in relation.split(":")[:2]]
```

Comment:
Single point of truth: what makes up the relevant parts of a relation is defined in class method `fetch_relation_parts`. Here that 'definition' is used inline. Should that definition ever change you need to update it in two places. Probably better to use:

```python
related_dataset, related_table = [to_snake_case(part) for part in RelationMaker.fetch_relation_parts(relation)]
```

Reply:
Valid point, updated!